jobs: TestTransientTxnErrors failed

[(jobs).TestTransientTxnErrors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2792767&tab=buildLog) on [master@ee9f47b9ec9476a693464e2dcd09a01bf9d39ad2](https://github.com/cockroachdb/cockroach/commits/ee9f47b9ec9476a693464e2dcd09a01bf9d39ad2):

```
=== RUN   TestTransientTxnErrors
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTransientTxnErrors660145572
    test_log_scope.go:74: use -show-logs to present logs inline
    job_scheduler_test.go:637: 
        	Error Trace:	job_scheduler_test.go:637
        	            				ctxgroup.go:175
        	            				ctxgroup.go:166
        	            				errgroup.go:57
        	            				asm_amd64.s:1374
        	Error:      	Received unexpected error:
        	            	savepoint error for schedule 642459219885555713: TransactionRetryWithProtoRefreshError: WriteTooOldError: write at timestamp 1616133397.549240657,0 too old; wrote at 1616133397.690849489,1: "unnamed" meta={id=064c6582 pri=0.00887313 epo=0 ts=1616133397.690849489,1 min=1616133397.549240657,0 seq=3} lock=true stat=PENDING rts=1616133397.549240657,0 wto=false gul=1616133398.049240657,0
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.(*jobScheduler).executeSchedules
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/job_scheduler.go:316
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.StartJobSchedulerDaemon.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/job_scheduler.go:471
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.TestTransientTxnErrors.func1.1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/job_scheduler_test.go:639
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:772
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:828
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:771
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.TestTransientTxnErrors.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/job_scheduler_test.go:638
        	            	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GroupWorkers.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:175
        	            	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57
        	            	  | runtime.goexit
        	            	  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
        	            	Wraps: (2) savepoint error for schedule 642459219885555713
        	            	Wraps: (3) TransactionRetryWithProtoRefreshError: WriteTooOldError: write at timestamp 1616133397.549240657,0 too old; wrote at 1616133397.690849489,1: "unnamed" meta={id=064c6582 pri=0.00887313 epo=0 ts=1616133397.690849489,1 min=1616133397.549240657,0 seq=3} lock=true stat=PENDING rts=1616133397.549240657,0 wto=false gul=1616133398.049240657,0
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *jobs.savePointError
        	Test:       	TestTransientTxnErrors
E210319 05:56:42.364837 27153 jobs/adopt.go:260  [n1] 1  job 642459246121648129: adoption completed with error node unavailable; try another peer
    job_scheduler_test.go:648: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTransientTxnErrors660145572
--- FAIL: TestTransientTxnErrors (8.32s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestTransientTxnErrors PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestTransientTxnErrors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(jobs).TestTransientTxnErrors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2803949&tab=buildLog) on [master@8ef4f671c4f008c5b871486fa2e780657a53e9e3](https://github.com/cockroachdb/cockroach/commits/8ef4f671c4f008c5b871486fa2e780657a53e9e3):

```
=== RUN   TestTransientTxnErrors
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTransientTxnErrors992852154
    test_log_scope.go:74: use -show-logs to present logs inline
    job_scheduler_test.go:637: 
        	Error Trace:	job_scheduler_test.go:637
        	            				ctxgroup.go:175
        	            				ctxgroup.go:166
        	            				errgroup.go:57
        	            				asm_amd64.s:1374
        	Error:      	Received unexpected error:
        	            	savepoint error for schedule 643586107315191809: TransactionRetryWithProtoRefreshError: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "unnamed" meta={id=03a6a347 key=/Table/52/1 pri=0.02528336 epo=0 ts=1616477295.615755305,0 min=1616477295.615755305,0 seq=28} lock=true stat=ABORTED rts=1616477295.615755305,0 wto=false gul=1616477296.115755305,0
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.(*jobScheduler).executeSchedules
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/job_scheduler.go:316
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.StartJobSchedulerDaemon.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/job_scheduler.go:471
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.TestTransientTxnErrors.func1.1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/job_scheduler_test.go:639
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:772
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:828
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:771
        	            	  | github.com/cockroachdb/cockroach/pkg/jobs.TestTransientTxnErrors.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/job_scheduler_test.go:638
        	            	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GroupWorkers.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:175
        	            	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:166
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57
        	            	  | runtime.goexit
        	            	  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
        	            	Wraps: (2) savepoint error for schedule 643586107315191809
        	            	Wraps: (3) TransactionRetryWithProtoRefreshError: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "unnamed" meta={id=03a6a347 key=/Table/52/1 pri=0.02528336 epo=0 ts=1616477295.615755305,0 min=1616477295.615755305,0 seq=28} lock=true stat=ABORTED rts=1616477295.615755305,0 wto=false gul=1616477296.115755305,0
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *jobs.savePointError
        	Test:       	TestTransientTxnErrors
    job_scheduler_test.go:648: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTransientTxnErrors992852154
--- FAIL: TestTransientTxnErrors (18.62s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestTransientTxnErrors PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestTransientTxnErrors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/jobs/job_scheduler_test.go
+++ b/pkg/jobs/job_scheduler_test.go
@@ -24,10 +24,10 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/scheduledjobs"
 	"github.com/cockroachdb/cockroach/pkg/security"
 	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -598,51 +598,124 @@ func TestJobSchedulerDaemonUsesSystemTables(t *testing.T) {
 	})
 }
 
+type txnConflictExecutor struct {
+	beforeUpdate, proceed chan struct{}
+}
+
+func (e *txnConflictExecutor) ExecuteJob(
+	ctx context.Context,
+	cfg *scheduledjobs.JobExecutionConfig,
+	env scheduledjobs.JobSchedulerEnv,
+	schedule *ScheduledJob,
+	txn *kv.Txn,
+) error {
+	// Read number of rows -- this count will be used when updating
+	// a single row in the table.
+	row, err := cfg.InternalExecutor.QueryRow(
+		ctx, "txn-executor", txn, "SELECT count(*) FROM defaultdb.foo")
+	if err != nil {
+		return err
+	}
+	cnt := int(tree.MustBeDInt(row[0]))
+	if e.beforeUpdate != nil {
+		// Wait to be signaled.
+		e.beforeUpdate <- struct{}{}
+		<-e.proceed
+		e.beforeUpdate = nil
+		e.proceed = nil
+	}
+
+	// Try updating.
+	_, err = cfg.InternalExecutor.Exec(
+		ctx, "txn-executor", txn, "UPDATE defaultdb.foo SET b=b+$1 WHERE a=1", cnt)
+	return err
+}
+
+func (e *txnConflictExecutor) NotifyJobTermination(
+	ctx context.Context,
+	jobID jobspb.JobID,
+	jobStatus Status,
+	details jobspb.Details,
+	env scheduledjobs.JobSchedulerEnv,
+	schedule *ScheduledJob,
+	ex sqlutil.InternalExecutor,
+	txn *kv.Txn,
+) error {
+	return nil
+}
+
+func (e *txnConflictExecutor) Metrics() metric.Struct {
+	return nil
+}
+
+var _ ScheduledJobExecutor = (*txnConflictExecutor)(nil)
+
 func TestTransientTxnErrors(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	skip.UnderRace(t, "takes >1min under race")
 
 	h, cleanup := newTestHelper(t)
 	defer cleanup()
 	ctx := context.Background()
 
-	h.sqlDB.Exec(t, "CREATE TABLE defaultdb.foo(a int primary key, b timestamp not null)")
+	h.sqlDB.Exec(t, `
+CREATE TABLE defaultdb.foo(a int primary key, b int);
+INSERT INTO defaultdb.foo VALUES(1, 1)
+`)
 
-	// Setup 10 schedules updating defaultdb.foo timestamp.
-	for i := 0; i < 10; i++ {
-		schedule := NewScheduledJob(h.env)
-		schedule.SetScheduleLabel(fmt.Sprintf("test schedule: %d", i))
-		schedule.SetOwner(security.TestUserName())
-		require.NoError(t, schedule.SetSchedule("*/1 * * * *"))
-		any, err := types.MarshalAny(&jobspb.SqlStatementExecutionArg{
-			Statement: fmt.Sprintf("UPSERT INTO defaultdb.foo (a, b) VALUES (%d, now())", i),
-		})
-		require.NoError(t, err)
-		schedule.SetExecutionDetails(InlineExecutorName, jobspb.ExecutionArguments{Args: any})
-		require.NoError(t, schedule.Create(
-			ctx, h.cfg.InternalExecutor, nil))
+	const execName = "test-executor"
+	ex := &txnConflictExecutor{
+		beforeUpdate: make(chan struct{}),
+		proceed:      make(chan struct{}),
 	}
+	defer registerScopedScheduledJobExecutor(execName, ex)()
 
-	// Setup numConcurrent workers, each executing maxExec executeSchedule calls.
-	const maxExec = 100
-	const numConcurrent = 3
+	// Setup schedule with our test executor.
+	schedule := NewScheduledJob(h.env)
+	schedule.SetScheduleLabel("test schedule")
+	schedule.SetOwner(security.TestUserName())
+	nextRun := h.env.Now().Add(time.Hour)
+	schedule.SetNextRun(nextRun)
+	schedule.SetExecutionDetails(execName, jobspb.ExecutionArguments{})
+	require.NoError(t, schedule.Create(
+		ctx, h.cfg.InternalExecutor, nil))
+
+	// Execute schedule on another thread.
+	g := ctxgroup.WithContext(context.Background())
+	g.Go(func() error {
+		return h.cfg.DB.Txn(context.Background(),
+			func(ctx context.Context, txn *kv.Txn) error {
+				err := h.execSchedules(ctx, allSchedules, txn)
+				return err
+			},
+		)
+	})
+
 	require.NoError(t,
-		ctxgroup.GroupWorkers(context.Background(), numConcurrent, func(ctx context.Context, _ int) error {
-			ticker := time.NewTicker(time.Millisecond)
-			numExecs := 0
-			for range ticker.C {
-				h.env.AdvanceTime(time.Minute)
-				// Transaction retry errors should never bubble up.
-				require.NoError(t,
-					h.cfg.DB.Txn(context.Background(), func(ctx context.Context, txn *kv.Txn) error {
-						return h.execSchedules(ctx, allSchedules, txn)
-					}))
-				numExecs++
-				if numExecs == maxExec {
-					return nil
-				}
+		h.cfg.DB.Txn(context.Background(), func(ctx context.Context, txn *kv.Txn) error {
+			// Let schedule start running, and wait for it to be ready to update.
+			h.env.SetTime(nextRun.Add(time.Second))
+			<-ex.beforeUpdate
+
+			// Before we let schedule proceed, update the number of rows in the table.
+			// This should cause transaction in schedule to restart, but we don't
+			// expect to see any errors in the schedule status.
+			if _, err := h.cfg.InternalExecutor.Exec(ctx, "update-a", txn,
+				`UPDATE defaultdb.foo SET b=3 WHERE a=1`); err != nil {
+				return err
 			}
+			if _, err := h.cfg.InternalExecutor.Exec(ctx, "add-row", txn,
+				`INSERT INTO defaultdb.foo VALUES (123, 123)`); err != nil {
+				return err
+			}
+			ex.proceed <- struct{}{}
 			return nil
 		}))
+
+	err := g.Wait()
+	require.True(t, errors.HasType(err, &savePointError{}))
+
+	// Reload schedule -- verify it doesn't have any errors in its status.
+	updated := h.loadSchedule(t, schedule.ScheduleID())
+	require.Equal(t, "", updated.ScheduleStatus())
 }

--- a/pkg/jobs/scheduled_job_executor.go
+++ b/pkg/jobs/scheduled_job_executor.go
@@ -55,28 +55,33 @@ type ScheduledJobExecutor interface {
 // ScheduledJobExecutorFactory is a callback to create a ScheduledJobExecutor.
 type ScheduledJobExecutorFactory = func() (ScheduledJobExecutor, error)
 
-var registeredExecutorFactories = make(map[string]ScheduledJobExecutorFactory)
+var executorRegistry struct {
+	syncutil.Mutex
+	factories map[string]ScheduledJobExecutorFactory
+	executors map[string]ScheduledJobExecutor
+}
 
 // RegisterScheduledJobExecutorFactory registers callback for creating ScheduledJobExecutor
 // with the specified name.
 func RegisterScheduledJobExecutorFactory(name string, factory ScheduledJobExecutorFactory) {
-	if _, ok := registeredExecutorFactories[name]; ok {
+	executorRegistry.Lock()
+	defer executorRegistry.Unlock()
+	if executorRegistry.factories == nil {
+		executorRegistry.factories = make(map[string]ScheduledJobExecutorFactory)
+	}
+
+	if _, ok := executorRegistry.factories[name]; ok {
 		panic("executor " + name + " already registered")
 	}
-	registeredExecutorFactories[name] = factory
+	executorRegistry.factories[name] = factory
 }
 
 // newScheduledJobExecutor creates new instance of ScheduledJobExecutor.
-func newScheduledJobExecutor(name string) (ScheduledJobExecutor, error) {
-	if factory, ok := registeredExecutorFactories[name]; ok {
+func newScheduledJobExecutorLocked(name string) (ScheduledJobExecutor, error) {
+	if factory, ok := executorRegistry.factories[name]; ok {
 		return factory()
 	}
 	return nil, errors.Newf("executor %q is not registered", name)
-}
-
-var executorRegistry struct {
-	syncutil.Mutex
-	executors map[string]ScheduledJobExecutor
 }
 
 // GetScheduledJobExecutor returns a singleton instance of
@@ -90,7 +95,7 @@ func GetScheduledJobExecutor(name string) (ScheduledJobExecutor, bool, error) {
 	if ex, ok := executorRegistry.executors[name]; ok {
 		return ex, false, nil
 	}
-	ex, err := newScheduledJobExecutor(name)
+	ex, err := newScheduledJobExecutorLocked(name)
 	if err != nil {
 		return nil, false, err
 	}

--- a/pkg/jobs/scheduled_job_executor_test.go
+++ b/pkg/jobs/scheduled_job_executor_test.go
@@ -72,9 +72,10 @@ func TestScheduledJobExecutorRegistration(t *testing.T) {
 	instance := newStatusTrackingExecutor()
 	defer registerScopedScheduledJobExecutor(executorName, instance)()
 
-	registered, err := newScheduledJobExecutor(executorName)
+	registered, created, err := GetScheduledJobExecutor(executorName)
 	require.NoError(t, err)
 	require.Equal(t, instance, registered)
+	require.True(t, created)
 }
 
 func TestJobTerminationNotification(t *testing.T) {

--- a/pkg/jobs/testutils_test.go
+++ b/pkg/jobs/testutils_test.go
@@ -149,7 +149,10 @@ func registerScopedScheduledJobExecutor(name string, ex ScheduledJobExecutor) fu
 			return ex, nil
 		})
 	return func() {
-		delete(registeredExecutorFactories, name)
+		executorRegistry.Lock()
+		defer executorRegistry.Unlock()
+		delete(executorRegistry.factories, name)
+		delete(executorRegistry.executors, name)
 	}
 }
 

